addons/mrp/report/mrp_report_mo_overview.py

```diff
@@ -191,22 +191,22 @@ def _get_mo_summary(self, production, components, current_mo_cost, current_real_
     def _get_unit_cost(self, move):
         return move.product_id.uom_id._compute_price(move.product_id.standard_price, move.product_uom)
 
-    def _format_state(self, record):
-        if record._name != 'mrp.production' or record.state not in ('draft', 'confirmed') or not record.move_raw_ids:
+    def _format_state(self, record, components=False):
+        """ For MOs, provide a custom state based on the demand vs quantities available for components.
+        All other records types will provide their standard state type.
+        :param dict components: components in the structure provided by `_get_components_data` 
+        :return: string to be used as custom state
+        """
+        if record._name != 'mrp.production' or record.state not in ('draft', 'confirmed') or not components:
             return dict(record._fields['state']._description_selection(self.env)).get(record.state)
-        components_qty_to_produce = defaultdict(float)
-        components_qty_reserved = defaultdict(float)
-        for move in record.move_raw_ids:
-            if move.product_id.detailed_type != 'product':
-                continue
-            components_qty_to_produce[move.product_id] += move.product_qty
-            components_qty_reserved[move.product_id] += move.product_uom._compute_quantity(move.quantity, move.product_id.uom_id)
         producible_qty = record.product_qty
-        for product_id, comp_qty_to_produce in components_qty_to_produce.items():
-            if float_is_zero(comp_qty_to_produce, precision_rounding=product_id.uom_id.rounding):
+        for component in components:
+            component = component["summary"]
+            # skip consumables + products with nothing to consume
+            if isinstance(component["quantity_free"], bool) or float_is_zero(component["quantity"], precision_rounding=component["uom_precision"]):
                 continue
             comp_producible_qty = float_round(
-                record.product_qty * (components_qty_reserved[product_id] + product_id.free_qty) / comp_qty_to_produce,
+                record.product_qty * (component["quantity_reserved"] + component["quantity_free"]) / component["quantity"],
```

Comment:
The previous loop (to know how many components are needed in total by product) made sense, if you have a disparity between quantities required for the same component.

e.g. the following case :
![image](https://github.com/odoo/odoo/assets/18055894/ab82ef1b-f23a-4f36-870a-9bb7d4d4d3f5)
The first move will get the full 10 available in its reservation, but the second one will have 0, meaning the second iteration will return 'Not Ready'.
But it's incorrect, as there are enough components for 1 finished product.

By changing that, it should also avoid counting the `free_qty` twice if there is some quantity available for a duplicated component. :thinking:

Reply:
D'oh, I knew there was a reason why I did it that way in the first place